### DISCUSSION TITLE
addRedactAnnot - Inaccurate rectangle position

### DISCUSSION THREAD

**Discussion by lchade** [pymupdf/PyMuPDF#1694] [Looking for help]
Hello,

We're using addRedactAnnot to annotate keywords with red rectangle on pdf pages, with this code statement:

`page.addRedactAnnot(area, text=sometext, cross_out=False, fill=(0.75, 0.8, 0.95), text_color=(1,1,1))`

In a pdf file we've got this result:
![image](https://user-images.githubusercontent.com/44476697/165590034-5ebaee9e-7da1-44f4-8d84-b10179166637.png)

On this page, "Mr. Reza Nakhiavani" and "Ms. Fatemeh Javaheri" are supposed to be annotated. We printed the annotated words to the console and saw the code actually manages to capture the correct text, but it just doesn't show correctly on the pdf file.

Is it because the pdf file itself messes something up? Any possibility to fix this? Thanks.

**Comment by JorjMcKie** [comment]
How did you compute the redact rectangles?
If via a PyMuPDF search, then what still ay go wrong is a sloppy "geometry change" done by the PDF creator - see documentation under "misplaced items".
Heal this by doing a `page.clean_contents()` once before trying any insertion.

**Reply by jeff-ywang** [reply]
Thanks for the quick response. Are you suggesting calling page.clean_contents() before each insertion? The documentation says "before you do your first item insertion.". [link](https://pymupdf.readthedocs.io/en/latest/faq.html#misplaced-item-insertions-on-pdf-pages)

**Reply by JorjMcKie** [reply]
only once!
The method cleans the resp. source code - once ok, you are good to go.

**Reply by jeff-ywang** [reply]
Thanks for the clarification.

**Reply by lchade** [reply]
> only once! The method cleans the resp. source code - once ok, you are good to go.

Hello,

We've added a `page.clean_contents()` and let it run before the annotation statement but the issue remains. This is the test file:
[annotate test page.pdf](https://github.com/pymupdf/PyMuPDF/files/8585330/annotate.test.page.pdf)
 
Would you try on your side, searching for and annotating "Mr. Reza Nakhjavani" and "Ms. Fatemeh Javaheri" on the page and see if the annotation position is same as the previous screenshot we posted?

Thanks

**Reply by lchade** [reply]
> > only once! The method cleans the resp. source code - once ok, you are good to go.
> 
> Hello,
> 
> We've added a `page.clean_contents()` and let it run before the annotation statement but the issue remains. This is the test file: [annotate test page.pdf](https://github.com/pymupdf/PyMuPDF/files/8585330/annotate.test.page.pdf)
> 
> Would you try on your side, searching for and annotating "Mr. Reza Nakhjavani" and "Ms. Fatemeh Javaheri" on the page and see if the annotation position is same as the previous screenshot we posted?
> 
> Thanks

@JorjMcKie I am sorry to have to ask for your assistance again. Have you tried the test? Is there anything else (more test files, our code snippet, etc.) we can provide to help the test?

**Reply by JorjMcKie** [reply]
I did try it out. It is a weird case indeed. Found no explanation yet.
I am also afraid I cannot spend much time to pursue the reason.

**Reply by lchade** [reply]
Ok no problem, thanks for maintaining and supporting this
